### PR TITLE
VV-2647 Split the .h file into a .h file and a .cpp file. Remove all …

### PR DESCRIPTION
**Pull request by MarkDeBrabander1976**
**Changes:**
Split the header file into a header file and cpp file.
Remove all the inline keywords for all functions, they are not needed anymore after the split in .h and .cpp file.
Dynamically create the display buffers only once.
Add a define for the maximum amount of pixels

**Tested:**
Functionally tested on 6 times 64x32 matrices

### CODE REVIEW

**Review by ElmerAbel — APPROVED** [review]

File: PxMatrix.cpp [hunk 1]
```diff
@@ -0,0 +1,955 @@
...
+
+  if (height > PxMATRIX_MAX_HEIGHT) {
+#ifdef DEBUG_ESP_PORT
+    DEBUG_ESP_PORT.print("[PxMatrix] Height larger than PxMATRIX_MAX_HEIGHT.\n");
+#endif
+  }
+
+  if (width * height > PxMATRIX_MAX_PIXELS) {
+#ifdef DEBUG_ESP_PORT
+    DEBUG_ESP_PORT.print("[PxMatrix] Amount of needed pixelse bigger than PxMATRIX_MAX_PIXELS.\n");
```
Comment (rbnlgrw) [inline PxMatrix.cpp]: typo: pixelse

File: PxMatrix.cpp [hunk 2]
```diff
@@ -0,0 +1,955 @@
...
+#ifdef PxMATRIX_double_buffer
+  clearDisplay(1);
+#endif
+  setColorDepth(PxMATRIX_MAX_COLOR_DEPTH);
+}
+
+#ifdef ESP32
+void PxMATRIX::fm612xWriteRegister(uint16_t reg_value, uint8_t reg_position) {
+  spi_t* spi = SPI.bus();
+  // reg_value = 0x1234;  debug
```
Comment (rbnlgrw) [inline PxMatrix.cpp]: commented out code

File: PxMatrix.cpp [hunk 3]
```diff
@@ -0,0 +1,955 @@
...
+  for (int i = 0; i < 47; i++) SPI_2BYTE(reg_value);
+
+  spiSimpleTransaction(spi);
+
+  spi->dev->mosi_dlen.usr_mosi_dbitlen = 16 - reg_position - 1;
+  spi->dev->miso_dlen.usr_miso_dbitlen = 0;
+  spi->dev->data_buf[0] = reg_value >> 8;
+  spi->dev->cmd.usr = 1;
+  while (spi->dev->cmd.usr)
+    ;
```
Comment (rbnlgrw) [inline PxMatrix.cpp:108]: potentially infinite while loop, it'd be better if there was a timeout or some kind of error state checking. If this gets stuck it sucks to debug. There are a couple more further down.

File: PxMatrix.cpp [hunk 4]
```diff
@@ -0,0 +1,955 @@
...
+  GPIO_REG_SET(1 << _LATCH_PIN);
+
+  spi->dev->mosi_dlen.usr_mosi_dbitlen = (reg_position - 8) - 1;
+  spi->dev->data_buf[0] = reg_value >> (reg_position - 8);
+  spi->dev->cmd.usr = 1;
+  while (spi->dev->cmd.usr)
+    ;
+  spiEndTransaction(spi);
+
+  SPI_BYTE(reg_value & 0xff);
```
Comment (rbnlgrw) [inline PxMatrix.cpp:119]: weird way of casting a 2-byte value to a 1-byte value, i like how concise it is but it's not "correct"

File: PxMatrix.cpp [hunk 5]
```diff
@@ -0,0 +1,955 @@
...
+
+  GPIO_REG_CLEAR(1 << _LATCH_PIN);
+}
+#else
```
Comment (rbnlgrw) [inline PxMatrix.cpp:123]: We're never gonna use this else, right? Since we only use ESP32.

File: PxMatrix.cpp [hunk 6]
```diff
@@ -0,0 +1,955 @@
...
+    b12a = 0b0111100011111111;           // 亮度: low
+    // uint16_t b12b=0b0111100000111111;
+    // uint16_t b12c=0b0111111111111111;
+    // uint16_t b12d=0b0111100000111111;
+
+    uint16_t b13a = 0b0000000001000000;
+    // uint16_t b13b=0b0000000001000000;
+    // uint16_t b13c=0b0000000001000000;
+    // uint16_t b13d=0b0000000001000000;
+
```
Comment (rbnlgrw) [inline PxMatrix.cpp:200]: Bunch of commented out code with no specified reason

File: PxMatrix.cpp [hunk 7]
```diff
@@ -0,0 +1,955 @@
...
+void PxMATRIX::setMuxDelay(uint8_t mux_delay_A, uint8_t mux_delay_B, uint8_t mux_delay_C, uint8_t mux_delay_D,
+                           uint8_t mux_delay_E) {
+  _mux_delay_A = mux_delay_A;
+  _mux_delay_B = mux_delay_B;
+  _mux_delay_C = mux_delay_C;
+  _mux_delay_D = mux_delay_D;
+  _mux_delay_E = mux_delay_E;
+}
+
+void PxMATRIX::setScanPattern(scan_patterns scan_pattern) { _scan_pattern = scan_pattern; }
```
Comment (rbnlgrw) [inline PxMatrix.cpp:239]: oneliners like this probably violate our code style. i don't really mind though.

File: PxMatrix.cpp [hunk 8]
```diff
@@ -0,0 +1,955 @@
...
+    total_offset_r = _row_offset[y] - in_row_byte_offset -
+                     _panel_width_bytes *
+                         (_row_sets_per_buffer * (_panels_width * which_buffer + which_panel) + vert_index_in_buffer);
+  }
+
+  uint8_t bit_select = x % 8;
+  // Some panels have a byte wise row-changing scanning pattern and/or a bit changing pattern that will be taken care of
+  // here.
+  if ((y % (_row_pattern * 2)) < _row_pattern) {
+    // Variant of ZIGZAG pattern with bit oder reversed on lower part (starts on upper part)
```
Comment (rbnlgrw) [inline PxMatrix.cpp:425]: typo: bit oder

File: PxMatrix.cpp [hunk 9]
```diff
@@ -0,0 +1,955 @@
...
+#ifdef PxMATRIX_double_buffer
+  fillMatrixBuffer(x, y, r, g, b, !_active_buffer);
+#else
+  fillMatrixBuffer(x, y, r, g, b, false);
+#endif
+}
+
+// the most basic function, get a single pixel
+uint8_t PxMATRIX::getPixel(int8_t x, int8_t y) {
+  return (0);  // PxMATRIX_buffer[x+ (y/8)*LCDWIDTH] >> (y%8)) & 0x1;
```
Comment (rbnlgrw) [inline PxMatrix.cpp:508]: Why???

File: PxMatrix.cpp [hunk 10]
```diff
@@ -0,0 +1,955 @@
...
+    digitalWrite(_C_PIN, LOW);
+  }
+  if (_row_pattern >= 16) {
+    pinMode(_D_PIN, OUTPUT);
+    digitalWrite(_D_PIN, LOW);
+  }
+  if (_row_pattern >= 32) {
+    pinMode(_E_PIN, OUTPUT);
+    digitalWrite(_E_PIN, LOW);
+  }
```
Comment (rbnlgrw) [inline PxMatrix.cpp:592]: could be tied together with "else if" to explicitly describe that they're alternative code branches, though i guess the compiler is smart enough to optimize it anyway.

File: PxMatrix.cpp [hunk 11]
```diff
@@ -0,0 +1,955 @@
...
+      if (_mux_delay_D) delayMicroseconds(_mux_delay_D);
+    }
+
+    if (_row_pattern >= 32) {
+      if (value & 0x10)
+        digitalWrite(_E_PIN, HIGH);
+      else
+        digitalWrite(_E_PIN, LOW);
+      if (_mux_delay_E) delayMicroseconds(_mux_delay_E);
+    }
```
Comment (rbnlgrw) [inline PxMatrix.cpp:635]: same as above, these are a bunch of implicit branches in code path that should technically be if .. else if ..

File: PxMatrix.cpp [hunk 12]
```diff
@@ -0,0 +1,955 @@
...
+    if (value == 2)
+      digitalWrite(_C_PIN, LOW);
+    else
+      digitalWrite(_C_PIN, HIGH);
+
+    if (value == 3)
+      digitalWrite(_D_PIN, LOW);
+    else
+      digitalWrite(_D_PIN, HIGH);
+  }
```
Comment (rbnlgrw) [inline PxMatrix.cpp:658]: Same here, seems like the author is allergic to else if ladders, i'll stop pointing them out

File: PxMatrix.cpp [hunk 13]
```diff
@@ -0,0 +1,955 @@
...
+}
+
+void PxMATRIX::latch(uint16_t show_time) {
+  if (_driver_chip == SHIFT) {
+    // digitalWrite(_OE_PIN,0); // <<< remove this
```
Comment (rbnlgrw) [inline PxMatrix.cpp:663]: Yes, definitely remove this :joy:

File: PxMatrix.cpp [hunk 14]
```diff
@@ -0,0 +1,955 @@
...
+          // This pre-buffers the data for the first row pattern of the next _display_color
+          SPI_TRANSFER(&bufferp->Data[((_display_color + 1) % _color_depth)][0], _send_buffer_size);
+        }
+
+        while ((micros() - start_time) < latch_time) delayMicroseconds(1);
+        digitalWrite(_OE_PIN, HIGH);
+
+      } else {
+        set_mux(i);
+#ifdef __AVR__
```
Comment (rbnlgrw) [inline PxMatrix.cpp:746]: Not an architecture we use

**Review by rbnlgrw — APPROVED** [review]
The library is a bit of a mess, i suggest you discuss with Elmer whether it's worth fixing. Personally i don't think any of the problems are significant enough to warrant spending the time.

**Pull request merged**